### ISSUE TITLE
ElasticSearch 2.x leaving "deleted" dangling files open?

### ISSUE THREAD

**Issue by diranged** [elastic/elasticsearch#15716]
We've found a few nodes of our ElasticSearch 2.1.1 cluster in a state where there are a ton of "dangling" deleted files that ElasticSearch has not let go of, so the OS cannot let go of them. This is causing some major problems with shard-allocation-by-disk-utilization.

```
[root@ops-flume-es-data-useast1-156-i-30af8181:/mnt:130]# du -sch /mnt
2.5T	/mnt
2.5T	total
[root@ops-flume-es-data-useast1-156-i-30af8181:/mnt]# df -h /mnt
Filesystem      Size  Used Avail Use% Mounted on
/dev/md127      5.0T  4.4T  662G  88% /mnt
[root@ops-flume-es-data-useast1-156-i-30af8181:/mnt]# 
```

```
[root@ops-flume-es-data-useast1-156-i-30af8181:/mnt]# lsof | grep deleted
...
...
java        6130   6151 elasticsearch 2078r      REG              9,127          43     4198403 /mnt/elasticsearch/flume-elasticsearch-production_vpc-useast1/flume-elasticsearch-production_vpc-useast1/nodes/0/indices/events-2015-12-05/5/translog/translog-2.tlog (deleted)
java        6130   6151 elasticsearch 2079u      REG              9,127          43        2060 /mnt/elasticsearch/flume-elasticsearch-production_vpc-useast1/flume-elasticsearch-production_vpc-useast1/nodes/0/indices/events-2015-12-05/5/translog/translog-3.tlog (deleted)
java        6130   6151 elasticsearch 2080r      REG              9,127  2574004451 32227958823 /mnt/elasticsearch/flume-elasticsearch-production_vpc-useast1/flume-elasticsearch-production_vpc-useast1/nodes/0/indices/flume-2015-12-05/4/index/_36jq_Lucene50_0.doc (deleted)
java        6130   6151 elasticsearch 2081r      REG              9,127  9180063556 32227958821 /mnt/elasticsearch/flume-elasticsearch-production_vpc-useast1/flume-elasticsearch-production_vpc-useast1/nodes/0/indices/flume-2015-12-05/4/index/_36jq.fdt (deleted)
java        6130   6151 elasticsearch 2082r      REG              9,127  1083231791 32227958828 /mnt/elasticsearch/flume-elasticsearch-production_vpc-useast1/flume-elasticsearch-production_vpc-useast1/nodes/0/indices/flume-2015-12-05/4/index/_36jq_Lucene50_0.pos (deleted)
...
```

```
[root@ops-flume-es-data-useast1-156-i-30af8181:/mnt]# lsof | grep deleted | wc
  169786 1867204 45795894
```

**Comment by mikemccand** [comment]
Lucene will hold open file descriptors for deleted index files, when segment merging has happened but flush and/or refresh haven't happened.

Have you customized e.g. `refresh_interval` or any of the translog flush settings?

Do you ever invoke `optimize` (`forceMerge`)?

The still-open-but-deleted translog files are interesting ... @s1monw is that expected?

**Comment by clintongormley** [comment]
Hi @diranged - any further info re @mikemccand 's [questions](https://github.com/elastic/elasticsearch/issues/15716#issuecomment-168191515)

@bleskes any ideas about the still-open-but-deleted translog files?

**Comment by bleskes** [comment]
I looked at the code and the only place I could see where we might leak file references to the translog is if we had an exception while constructing the translog, followed by a successful opening of it. It also turns out that Simon already fixed it with  #15762 :grin: , released with 2.1.2

**Comment by clintongormley** [comment]
Looks like there's nothing left to do here. Closing
